streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -630,57 +631,188 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
+                         },
+                         { case (rightEx, leftFib) =>
+                           leftFib.interrupt *> rightEx
+                         }
+                       )
+                     //.debug("forkF.race")
+                   }
+                     .onExit(localScope.close(_))
+                     .fork
+          } yield fib
+      }
+
+      lazy val enqueueCh: ZChannel[Env1, OutErr, OutElem, OutDone, Nothing, Nothing, Any] = ZChannel
+        .readWithCause(
+          in => ZChannel.fromZIO(forkF(in).flatMap(queue.offer(_))) *> enqueueCh,
+          err =>
+            ZChannel.fromZIO(
+              failureSignal.failCause(err.map(Some(_))) *> queue.offer(Fiber.failCause(err.map(Some(_))))
+            ),
+          done => ZChannel.fromZIO(outDoneSignal.succeed(done) *> queue.offer(Fiber.fail(None)))
+        )
+
+      val enqueuer: ZIO[Env1 with Scope, Nothing, Fiber.Runtime[Nothing, Any]] = queueReader
+        .pipeTo(self)
+        .pipeTo(enqueueCh)
+        .runScoped
+        .forkScoped
+
+      lazy val readerCh: ZChannel[Any, Any, Any, Any, OutErr1, OutElem2, OutDone] =
+        ZChannel.unwrap {
+          val z0: URIO[Any, ZChannel[Any, Any, Any, Any, OutErr1, OutElem2, OutDone]] = queue.take
+            .flatMap(_.join)
+            .foldCause(
+              c =>
+                Cause
+                  .flipCauseOption(c)
+                  .map(ZChannel.refailCause(_))
+                  .getOrElse(ZChannel.fromZIO(outDoneSignal.await)),
+              out2 => ZChannel.write(out2) *> readerCh
             )
+          z0
+        }
+
+      val resCh: ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = ZChannel
+        .scoped[Env1](enqueuer)
+        .concatMapWith { enqueueFib =>
+          readerCh
+        }((_, o) => o, (o, _) => o)
+        .embedInput(input)
+
+      resCh
+    }
+
+    ZChannel.unwrap(z)
+  }
+
+  final def mapOutZIOParUnordered[Env1 <: Env, OutErr1 >: OutErr, OutElem2](n: Int, bufferSize: Int = 16)(
+    f: OutElem => ZIO[Env1, OutErr1, OutElem2]
+  )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = {
+    val z0: ZIO[Any, Nothing, ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone]] = for {
+      input      <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
+      queueReader = ZChannel.fromInput(input)
+      q          <- zio.Queue.bounded[Any](bufferSize)
+      permits    <- zio.Semaphore.make(n)
+    } yield {
+      def enqueue(a: OutElem): ZIO[Env1, Nothing, Unit] = ZIO.uninterruptibleMask { restore =>
+        for {
+          localScope <- zio.Scope.make
+          _          <- restore(permits.withPermitScoped.provideEnvironment(ZEnvironment(localScope)))
+          z1          = f(a)
+          offer = z1
+                    .foldCauseZIO(
+                      c => q.offer(QRes.failCause(c)),
+                      a2 => q.offer(a2)
+                    )
```

Comment:
not 100% sure how the compiler unfolds for-comprehensions but it makes sense in this case (the original code was a bit more complex and required a finer grained use of the `restore` function, but I eventually got to this simpler version which can be quite elegantly)

Reply:
The compiler won't re-write for-comprehensions, at least not at the moment. The Scala 3 compiler will be able to do that in the future.

You can reduce the number of flatmaps (since in ZIO a `map` is a `flatMap`), by adding `val offer = ???` in the block below as @paulpdaniels suggested